### PR TITLE
Fix/collate fn

### PR DESCRIPTION
**Pull request by zhu0619**
## Changelogs

- Add parameter `dataset`  to [PYGGraphTransformer.get_collate_fn](https://github.com/datamol-io/molfeat/blob/8272a6f67d99d8ad1fed078ca067cef3836c3d53/molfeat/trans/graph/adj.py#L746C26-L746C26)
- Add optional dependency `torch_geometric` in `env.yml` and `pyproject.toml`
---

_Checklist:_

- [x] _Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR.
- [x] _<s> Add tests to cover the fixed bug(s) or the new introduced feature(s) (if appropriate).</s>
- [x] _<s>Update the API documentation if a new function is added, or an existing one is deleted. Eventually consider making a new tutorial for new features.</s>
- [x] _Write concise and explanatory changelogs below._
- [x] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

_discussion related to that PR_

### CODE REVIEW

File: molfeat/trans/graph/adj.py [hunk 1]
```diff
@@ -736,14 +740,15 @@ def get_collate_fn(
...
+            dataset: The dataset from which to load the data and apply the collate function.
             follow_batch: Creates assignment batch vectors for each key in the list. (default: :obj:`None`)
             exclude_keys: Will exclude each key in the list. (default: :obj:`None`)
             return_pair: whether to return a pair of X,y or a databatch (default: :obj:`True`)
 
         Returns:
             Collated samples.
         """
-        collator = Collater(follow_batch=follow_batch, exclude_keys=exclude_keys)
+        collator = Collater(dataset=dataset, follow_batch=follow_batch, exclude_keys=exclude_keys)
```
Comment (maclandrol) [inline molfeat/trans/graph/adj.py:751]: Note that `dataset` can be None,  so it's in fact optional to provide (unless your dataset is an `OnDiskDataset` dataset), which the pyg documentation does not mention unfortunately

File: env.yml [hunk 2]
```diff
@@ -44,7 +44,8 @@ dependencies:
   - transformers
   - tokenizers <0.13.2
   - sentencepiece
-  - biotite # required for ESM models
+  - biotite # required for ESM model
+  - pytorch_geometric >=2.4.0
```
Comment (maclandrol) [inline env.yml:48]: Which machine have you tested this on ? I have a suspicion that some module might be required (e.g `pytorch_scatter`). 

Could you make sure that this notebook ([pyg integration](https://github.com/datamol-io/molfeat/blob/main/docs/tutorials/pyg_integration.ipynb)) works with your changes and adapt it if needed ?

File: molfeat/trans/graph/adj.py [hunk 3]
```diff
@@ -727,6 +730,7 @@ def transform(self, mols: List[Union[dm.Mol, str]], **kwargs):
 
     def get_collate_fn(
         self,
+        dataset: Union[Dataset, Sequence[BaseData], DatasetAdapter],
```
Comment (maclandrol) [inline molfeat/trans/graph/adj.py:733]: At first glance, I am convinced, this is likely going to fail if the package is not installed (which is the default behaviour with pip).

Please use something like:

```python
from typing import TYPE_CHECKING, Union, Sequence

if requires.check("torch_geometric"):
    from torch_geometric.data import Data
    from torch_geometric.loader.dataloader import Collater
    if TYPE_CHECKING:
         from torch_geometric.data import Dataset # should probably import as PygDataset
         from torch_geometric.data.data import BaseData
         from torch_geometric.data.datapipes import DatasetAdapter
    else:
         Dataset, BaseData, DatasetAdapter = None, None, None
         
```

Then use strings for the typing annotations. e.g `Sequence["BaseData"]`

**Review by maclandrol — COMMENTED** [review]
Thanks @zhu0619, please see comments

File: pyproject.toml [hunk 4]
```diff
@@ -76,6 +78,7 @@ all = [
     "fcd_torch",
     "nglview",
     "ipywidgets",
+    "pytorch_geometric"
```
Comment (maclandrol) [inline pyproject.toml:81]: please set version here

File: env.yml (hunk 2 above)
Comment (zhu0619) [reply env.yml:48]: I am testing on Mac m2. The tutorial runs with no issues. It seems pytorch_scatter is not necessary here.

**Comment by zhu0619** [comment]
@maclandrol Let me know if it's ok to be merged.

**Review by maclandrol — APPROVED** [review]

**Comment by maclandrol** [comment]
@zhu0619  feel free to merge and release after the CI is green.

**Pull request merged**